File: apps/mobile/lib/core/user_code.dart
/// 8 位用户 ID 工具（与 Web `user_code.ts` 对齐；兼容历史 10 位）。
library;

class UserCode {
  static const code8 = r'^\d{8}$';
  static const code10 = r'^\d{10}$';

  static bool isUserCode(String value) =>
      RegExp(code8).hasMatch(value) || RegExp(code10).hasMatch(value);

  static int _imul(int a, int b) {
    final r = (a * b) & 0xFFFFFFFF;
    return r > 0x7FFFFFFF ? r - 0x100000000 : r;
  }

  static String deviceIdToUserCode(String deviceId) {
    var hash = 0;
    for (var i = 0; i < deviceId.length; i++) {
      hash = _imul(31, hash) + deviceId.codeUnitAt(i);
      if (hash > 0x7FFFFFFF) hash -= 0x100000000;
    }
    final n = hash.abs() % 90000000 + 10000000;
    return n.toString();
  }
}

File: apps/web/lib/user_code.ts
/** 用户 ID 格式：新用户 8 位数字；兼容历史 10 位。 */

export const USER_CODE_LEN = 8;
export const USER_CODE_RE = /^\d{8}$/;
export const LEGACY_USER_CODE_RE = /^\d{10}$/;

export function isUserCode(value: string): boolean {
  return USER_CODE_RE.test(value) || LEGACY_USER_CODE_RE.test(value);
}

/** 由设备 ID 确定性生成 8 位用户码（与 App 端算法一致）。 */
export function deviceIdToUserCode(deviceId: string): string {
  let hash = 0;
  for (let i = 0; i < deviceId.length; i += 1) {
    hash = (Math.imul(31, hash) + deviceId.charCodeAt(i)) | 0;
  }
  const n = (Math.abs(hash) % 90_000_000) + 10_000_000;
  return String(n);
}
